### ISSUE TITLE
ERROR: IOError: readdir("/boot/efi"): permission denied (EACCES)

### ISSUE THREAD

**Issue by abassign** [yha/ScanDir.jl#4]
Hello,
I tried your procedure (congratulations for the execution speed compared to the system one, I hope it becomes official!) but like the official one there is a problem when, during the scan, a read-protected file is extracted.

```
julia> for (root, dirs, files) in ScanDir.walkdir("/") println(files) end
["bin", "lib", "lib32", "lib64", "libx32", "sbin", "swapfile"]
["System.map-5.11.0-17-generic", "System.map-5.11.0-18-generic", "config-5.11.0-17-generic", "config-5.11.0-18-generic", "initrd.img", "initrd.img-5.11.0-17-generic", "initrd.img-5.11.0-18-generic", "initrd.img.old", "memtest86+.bin", "memtest86+.elf", "memtest86+_multiboot.bin", "vmlinuz", "vmlinuz-5.11.0-17-generic", "vmlinuz-5.11.0-18-generic", "vmlinuz.old"]
ERROR: IOError: readdir("/boot/efi"): permission denied (EACCES)
Stacktrace:
 [1] check_channel_state
   @ ./channels.jl:170 [inlined]
 [2] take_unbuffered(c::Channel{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}})
   @ Base ./channels.jl:403
 [3] take!
   @ ./channels.jl:383 [inlined]
 [4] iterate(c::Channel{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}}, state::Nothing)
   @ Base ./channels.jl:465
 [5] top-level scope
   @ REPL[35]:1
```
Unfortunately I can't manage this type of error as it is generated at the Channel level. I think the best solution would be to insert a filter that excludes a directory and/or file by attribute, what do you think? It is feasible ?
Unfortunately walkdir is an excellent method for searching for files within a file system, but it does not have file/directory attribute exclusion filters.

What do you think about it ?

**Comment by yha** [comment]
> I think the best solution would be to insert a filter that excludes a directory and/or file by attribute, what do you think? It is feasible ?

You can filter results using the `prune` keyword argument. If you want to filter by permission in the pruning function, I think you would need to use `stat` and interpret the result "manually". I'm not sure if a nicer interface to the `stat` result is in the scope of this package.
It's probably easier in this case to use `onerror`:
```
julia> collect(ScanDir.walkdir("/etc/"; onerror=e->(showerror(stderr,e); println(stderr))))
IOError: readdir("/etc/polkit-1/localauthority"): permission denied (EACCES)
IOError: readdir("/etc/ssl/private"): permission denied (EACCES)
286-element Vector{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}}:
[...]
```

**Comment by abassign** [comment]
@yha 
Hi, In the meantime, thank you for the answer, but I found a more integrated solution in the code which could be the basis of an additional option.
I noticed that the error only occurs for directories defined like this:
```
jp = joinpath (root, dir.name)
Which have gperm == 0x00
```
And so I entered this code
```
function scandirtree(root="."; topdown=true, follow_symlinks=false, onerror=throw, prune=_->false)
    ...
        for dir in dirs
            jp = joinpath(root, dir.name)
            gperm(jp) != 0x00 && _scandirtree(chnl, jp)
        end
     ...
```
which made everything work perfectly.

At this point it might be useful to insert a parameter that defines to bypass the directories with gperm (or other flags) which are of value 0x00.

In fact, scandirtree (...) is a very useful function that allows you to operate excellently on the filesystem, but it becomes completely useless when it encounters a single problem that it cannot bypass.
It would be useful to do the same for hidden directories that we don't always want to scan, for which another flag that allows them to be bypassed would be welcome.

What do you think about it ?

I currently have it solved so here is the program I am developing with your modified code:

https://github.com/abassign/Photoscenary

Greetings
Adriano

**Comment by yha** [comment]
I don't see why you would want to do the check `gperm(jp) != 0x00` *inside* the scandir implementation. Wouldn't it be better to just user `onerror` to print out any read errors as warnings? Or at least do this `gperm` filtering inside `prune`?

> but it becomes completely useless when it encounters a single problem that it cannot bypass.

If you want "bypass" an error, just pass some appropriate function as `onerror`.

A problem with group permissions is just one possible reason for an error on reading the directory (and the particular case of gperm 0x00 is even more specific). I don't think it's a good idea to add a flag for every possible problem to the API. It's both less elegant and less robust then using `prune` and/or `onerror`.

It could be nice to have a utility function for "can I read this" that checks the user identity as well as `gperm`, `uperm` and `operm` as necessary, so you can use it in `prune`.

**Comment by abassign** [comment]
@yha 

The error function linked to channel is perfectly unrelated to the context. This means that when the error occurs it is not possible to know where it occurs, and therefore it is of little use.
But actually I don't even see that the _gperm(jp) == 0x00_ state is an error, it's just one of the possible filesystem states that says that specific file cannot be read.
If a node cannot be read it is obvious that _scandirtree cannot enter to read it, otherwise it violates the reason for the restriction generated by the filesystem. So that node is in fact a terminal node.
`gperm(jp)! = 0x00 && _scandirtree (chnl, jp)`
Activate the _scandirtree recursive function only if the node is not terminal.
In fact, it makes no sense to activate an error function, as it is not an error condition, it is simply a different termination condition of the recursive navigation function of that specific branch of the root.
The only thing I think it might be useful to put a switch between the parameters to activate the control if required.
Type this code:
`function scandirtree (root = "."; topdown = true, follow_symlinks = false, onerror = throw, prune = _-> false, permission_0_check = true)`
And then use this recursive code:

```
        for dir in dirs
            jp = joinpath (root, dir.name)
            permission_0_check ? gperm(jp)!= 0x00 && _scandirtree(chnl, jp) : _scandirtree(chnl, jp)
        end
```

Which allows you to choose one of the two ways, while avoiding raising an error. You can also raise an error, but in a channel an error blocks the whole process making it impossible to read a root that contains read-protected files and therefore, for example, _scandirtree **does not work on a Windows** home directory where there is always a certain number of read-protected files.

**In summary, the error that currently blocks the function if it encounters a known unreadable is not an error, but simply an incorrectly managed file system state. Therefore the check I inserted, blocking an action not allowed by the file-system, avoids the problem analyzed in this post.**

**Comment by yha** [comment]
> The error function linked to channel is perfectly unrelated to the context. This means that when the error occurs it is not possible to know where it occurs, and therefore it is of little use.

I'm not sure what that means. If I do 
```
foo(p) = collect(ScanDir.walkdir(p))

foo("/etc/")
```
I get
```
ERROR: LoadError: IOError: readdir("/etc/polkit-1/localauthority"): permission denied (EACCES)
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait
    @ ./task.jl:769 [inlined]
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_unbuffered(c::Channel{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}})
    @ Base ./channels.jl:405
  [5] take!
    @ ./channels.jl:383 [inlined]
  [6] iterate(c::Channel{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}}, state::Nothing)
    @ Base ./channels.jl:465
  [7] _collect(cont::UnitRange{Int64}, itr::Channel{NamedTuple{(:root, :dirs, :files), Tuple{String, Vector{String}, Vector{String}}}}, #unused#::Base.HasEltype, isz::Base.SizeUnknown)
    @ Base ./array.jl:614
  [8] collect
    @ ./array.jl:602 [inlined]
  [9] foo(p::String)
    @ Main ./Untitled-1:3
 [10] top-level scope
    @ Untitled-1:4
in expression starting at Untitled-1:4
```
The error message mentions a permission issue and names the offending directory. And the stack trace correctly points to `foo(p::String)` which is the user code location of the problem. How is this error of little use?

> In fact, it makes no sense to activate an error function, as it is not an error condition, it is simply a different termination condition of the recursive navigation function of that specific branch of the root.

I disagree. Treating directories where there is no read permission as leaves of the directory tree may be the desired behavior in some cases. In other cases, it would be an unexpected error condition. I think it makes more sense to be conservative and make it an error by default, with an API that allows either behavior.
So I agree it might be nice to have a utility for checking read permissions (so that it can be used in `prune`). I think this would be quite subtle to implement correctly on all platforms (`gperm(..) != 0x00` is certainly not the right implementation. It is neither a sufficient nor a necessary condition for the current process to have read permissions of the directory).

> for example, _scandirtree **does not work on a Windows** home directory where there is always a certain number of read-protected files.

Works for me: Doing `collect(ScanDir.walkdir(homedir(); onerror = e->(showerror(stderr,e); println(stderr))))`, I see three "permission denied" messages, and then it correctly returns the list of `(root, dirs, files)` named tuples of what it can read.

**Comment by abassign** [comment]
@yha 
thank you for the answer that clarified the problem and pointed out that there is already a method of handling errors in Julia within these types of loops.
Since I don't have to show users of the program that I'm developing Windows induced system errors, I removed the change I made, and inserted a simple function called: countDirError
This function simply counts the number of errors encountered and counts them.
In this way I keep track of the problem, but I do not "scare" the user of the program who on average considers Windows perfect and at the same time I avoid the function from crashing.

```
function countDirError()
    dirsWithErrors::Int = 0
    add(err) = dirsWithErrors += 1
    get() = dirsWithErrors
    () -> (add;get)
end

function updateFilesListTypeDDS(path::String=homedir())
    filesPath = Dict{Int64,TailGroupByIndex}()
    cde = countDirError()
    if ispath(path)
	for (root, dirs, files) in ScanDir.walkdir(path; onerror = e->(cde.add(e)))
            ...
```
At this point I consider the issue closed
